File: DoublyLinkedListCS/DoublyLinkedListCS/NodeContainer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DoublyLinkedListCS
{
    class NodeContainer
    {
        Node head;
        Node last;
        int size;

        void AddToEnd(Node n)
        {
            if (last == null)
            {
                head = n;
                last = n;
            }
            else
            {
                last.pNext = n;
                n.pPrev = last;
                last = n;
            }
            size++;
        }

        public void CreateList()
        {
            Node nodeA = new Node("A", 12); //A
            this.AddToEnd(nodeA);
             

            Node nodeB = new Node("B", 2111); //B
            this.AddToEnd(nodeB);
             

            Node nodeC = new Node("C", 2121);
            this.AddToEnd(nodeC);
             

            Node nodeD = new Node("D", 211);
            this.AddToEnd(nodeD);
             

            Node nodeE = new Node("E", 1);
            this.AddToEnd(nodeE);
             

            Node nodeF = new Node("F", 1121);
            this.AddToEnd(nodeF);
             

            Node nodeG = new Node("G", 221);
            this.AddToEnd(nodeG);
             

            Node nodeH = new Node("H", 1111);
            this.AddToEnd(nodeH);
             

            Node nodeI = new Node("I", 11);
            this.AddToEnd(nodeI);
             

            Node nodeJ = new Node("J", 1222);
            this.AddToEnd(nodeJ);
             

            Node nodeK = new Node("K", 212);
            this.AddToEnd(nodeK);
             

            Node nodeL = new Node("L", 1211);
            this.AddToEnd(nodeL);
             

            Node nodeM = new Node("M", 22);
            this.AddToEnd(nodeM);
             

            Node nodeN = new Node("N", 21);
            this.AddToEnd(nodeN);
             

            Node nodeO = new Node("O", 222);
            this.AddToEnd(nodeO);
             

            Node nodeP = new Node("P", 1221);
            this.AddToEnd(nodeP);
             

            Node nodeQ = new Node("Q", 2212);
            this.AddToEnd(nodeQ);
             

            Node nodeR = new Node("R", 121);
            this.AddToEnd(nodeR);
             

            Node nodeS = new Node("S", 111);
            this.AddToEnd(nodeS);
             

            Node nodeT = new Node("T", 2);
            this.AddToEnd(nodeT);
             

            Node nodeU = new Node("U", 112);
            this.AddToEnd(nodeU);
             

            Node nodeV = new Node("V", 1112);
            this.AddToEnd(nodeV);
             

            Node nodeW = new Node("W", 122);
            this.AddToEnd(nodeW);
             

            Node nodeX = new Node("X", 2112);
            this.AddToEnd(nodeX);
             

            Node nodeY = new Node("Y", 2122);
            this.AddToEnd(nodeY);
             

            Node nodeZ = new Node("Z", 2211);
            this.AddToEnd(nodeZ);
             
        }

        public void PrintList()
        {
            if (size > 0)
            {
                Node current = head;
                while (current != null)
                {
                    Console.WriteLine("Node int: {0} \t Node char: {1}", current.intRepr, current.charRepr);
                    current = current.pNext;
                }
                Console.WriteLine("List Size: {0}", size);
            }
            else
            {
                Console.WriteLine("Empty List");
            }
        }

        public void DeleteNode(string s)
        {
            if (head != null)
            {
                Node current = head;
                while (current != null)
                {
                    if (current.charRepr == s)
                    {
                        if (current == head && current.pNext == null)
                        {
                            // only one node
                            current = null;
                            head = null;
                        }
                        else if (current == head)
                        {
                            // the first node
                            head = current.pNext;
                            current.pNext.pPrev = null;
                            current = null;
                        }
                        else if (current.pNext == null)
                        {
                            // the last node
                            current.pPrev.pNext = null;
                            current = null;
                        }
                        else
                        {
                            // a middle node
                            current.pPrev.pNext = current.pNext;
                            current.pNext.pPrev = current.pPrev;
                            current = null;
                        }
                        Console.WriteLine("Deleted {0} from the list.", s);
                        size--;
                        break;
                    }
                    current = current.pNext;
                }
            }
        }

        public void SortList()
        {
            if (head != null && head.pNext != null)
            {
                Node min = head;
                Node current = min.pNext;
                Node newHead = null;
                Node newTail = null;
                
                while (current != null)
                {
                    if (current.intRepr < min.intRepr)
                    {
                        min = current;
                    }
                    current = current.pNext;
                }
                
                if (min == head)
                {
                    head = min.pNext;
                    min.pNext.pPrev = null;
                }
                else if (min.pNext == null)
                {
                    min.pPrev.pNext = null;
                }
                else
                {
                    min.pNext.pPrev = min.pPrev;
                    min.pPrev.pNext = min.pNext;
                }
                min.pNext = null;
                min.pPrev = null;
                newHead = min;
                newTail = min;
                min = head;
                while (min != null && head != null)
                {
                    min = head;
                    current = min;
                    while (current != null)
                    {
                        if (current.intRepr < min.intRepr)
                        {
                            min = current;
                        }
                        current = current.pNext;
                    }
                    
                    if (min == head && min.pNext == null)
                    {
                        head = null;
                    }
                    else if (min == head)
                    {
                        head = min.pNext;
                        min.pNext.pPrev = null;
                    }
                    else if (min.pNext == null)
                    {
                        min.pPrev.pNext = null;
                    }
                    else
                    {
                        min.pNext.pPrev = min.pPrev;
                        min.pPrev.pNext = min.pNext;
                    }
                    min.pNext = null;
                    min.pPrev = newTail;
                    newTail.pNext = min;
                    newTail = min;

                    current = min.pNext;
                }

                head = newHead;
            }
        }
    }
}

File: DoublyLinkedList/DoublyLinkedList/NodeContainer.cpp
#include "stdafx.h"
#include "NodeContainer.h"

NodeContainer::NodeContainer() {
	head = nullptr;
	last = nullptr;
	size = 0;
}

NodeContainer::NodeContainer(NodeContainer& n) {
	head = n.head;
	last = n.last;
	size = n.size;
}

NodeContainer& NodeContainer::operator=(NodeContainer& n) {
	head = n.head;
	last = n.last;
	size = n.size;

	return *this;
}

NodeContainer::~NodeContainer() {

}

void NodeContainer::add(Node* n) {
	if (head == nullptr) {
		head = n;
		last = n;
	}
	else {
		n->setNext(head);
		head->setPrev(n);
		head = n;
	}
	size++;
}

void NodeContainer::addToEnd(Node* n) {
	if (last == nullptr) {
		head = n;
		last = n;
	}
	else {
		last->setNext(n);
		n->setPrev(last);
		last = n;
	}
	size++;
}

void NodeContainer::deleteNode(char c) {
	Node* n = findNode(c);

	if (n == nullptr) {
		//do nothing
	}
	else {
		if (n->getPrev() != nullptr && n->getNext() != nullptr) {
			//node has a previous and a next
			n->getPrev()->setNext(n->getNext());
			n->getNext()->setPrev(n->getPrev());
		}
		else if (n->getPrev() != nullptr) {
			//node has a previous but not a next -> end of list
			n->getPrev()->setNext(nullptr);
			last = n->getPrev();
		}
		else if (n->getNext() != nullptr) {
			//node has a next but not a previous -> beginning of list
			n->getNext()->setPrev(nullptr);
			head = n->getNext();
		}
		else {
			//only node in list
			head = nullptr;
			last = nullptr;
		}
		printf("Deleted %c\n", n->getChar());
		delete n;
		size--;
	}
	
}

Node* NodeContainer::findNode(char c) {
	Node* current = this->head;
	while (current != nullptr) {
		if (current->getChar() == c) break;
		else current = current->getNext();
	}
	return current;
}

void NodeContainer::selectionSort() {
	Node* current = this->head;
	if (current && current->next)
	{
		Node* newHead = 0;
		Node* newTail = 0;
		Node* min = current;
		while (current != nullptr)
		{
			if (current->getInt() < min->getInt())
			{
				min = current;
			}
			current = current->next;
		}

		if (min->prev != nullptr && min->next != nullptr)
		{
			// not the first node
			// not the last node
			min->prev->next = min->next;
			min->next->prev = min->prev;
		}
		else if (min->prev == nullptr && min->next != nullptr)
		{
			// first node
			min->next->prev = 0;
			this->head = min->next;
		}
		else if (min->prev != nullptr && min->next == nullptr)
		{
			// last node
			min->prev->next = 0;
		}
		else
		{
			// only one node
			this->head = 0;
		}
		newHead = min;
		newTail = min;
		min = current = this->head;
		while (min)
		{

			while (current)
			{
				if (current->getInt() < min->getInt())
				{
					min = current;
				}
				current = current->next;
			}

			if (min->prev && min->next)
			{
				// not the first node
				// not the last node
				min->prev->next = min->next;
				min->next->prev = min->prev;
			}
			else if (!min->prev && min->next)
			{
				// first node
				min->next->prev = 0;
				this->head = min->next;
			}
			else if (min->prev && !min->next)
			{
				// last node
				min->prev->next = 0;
			}
			else
			{
				// only one node
				this->head = 0;
			}
			newTail->next = min;
			min->prev = newTail;
			newTail = min;
			min = current = this->head;
		}
		this->head = newHead;
	}
}

void NodeContainer::printList() {
	if (size > 0) {
		Node* current = head;
		while (current != nullptr) {
			printf("Node int: %d \t Node char: %c\n", current->getInt(), current->getChar());

			current = current->getNext();
		}
		printf("List size: %d \n\n", size);
	}
	else {
		printf("Empty list. \n\n");
	}
}

void NodeContainer::createList() {
	char ch = 65;
	Node* nodeA = new Node(ch, 12); //A
	this->addToEnd(nodeA);
	ch++;

	Node* nodeB = new Node(ch, 2111); //B
	this->addToEnd(nodeB);
	ch++;

	Node* nodeC = new Node(ch, 2121);
	this->addToEnd(nodeC);
	ch++;

	Node* nodeD = new Node(ch, 211);
	this->addToEnd(nodeD);
	ch++;

	Node* nodeE = new Node(ch, 1);
	this->addToEnd(nodeE);
	ch++;

	Node* nodeF = new Node(ch, 1121);
	this->addToEnd(nodeF);
	ch++;

	Node* nodeG = new Node(ch, 221);
	this->addToEnd(nodeG);
	ch++;

	Node* nodeH = new Node(ch, 1111);
	this->addToEnd(nodeH);
	ch++;

	Node* nodeI = new Node(ch, 11);
	this->addToEnd(nodeI);
	ch++;

	Node* nodeJ = new Node(ch, 1222);
	this->addToEnd(nodeJ);
	ch++;

	Node* nodeK = new Node(ch, 212);
	this->addToEnd(nodeK);
	ch++;

	Node* nodeL = new Node(ch, 1211);
	this->addToEnd(nodeL);
	ch++;

	Node* nodeM = new Node(ch, 22);
	this->addToEnd(nodeM);
	ch++;

	Node* nodeN = new Node(ch, 21);
	this->addToEnd(nodeN);
	ch++;

	Node* nodeO = new Node(ch, 222);
	this->addToEnd(nodeO);
	ch++;

	Node* nodeP = new Node(ch, 1221);
	this->addToEnd(nodeP);
	ch++;

	Node* nodeQ = new Node(ch, 2212);
	this->addToEnd(nodeQ);
	ch++;

	Node* nodeR = new Node(ch, 121);
	this->addToEnd(nodeR);
	ch++;

	Node* nodeS = new Node(ch, 111);
	this->addToEnd(nodeS);
	ch++;

	Node* nodeT = new Node(ch, 2);
	this->addToEnd(nodeT);
	ch++;

	Node* nodeU = new Node(ch, 112);
	this->addToEnd(nodeU);
	ch++;

	Node* nodeV = new Node(ch, 1112);
	this->addToEnd(nodeV);
	ch++;

	Node* nodeW = new Node(ch, 122);
	this->addToEnd(nodeW);
	ch++;

	Node* nodeX = new Node(ch, 2112);
	this->addToEnd(nodeX);
	ch++;

	Node* nodeY = new Node(ch, 2122);
	this->addToEnd(nodeY);
	ch++;

	Node* nodeZ = new Node(ch, 2211);
	this->addToEnd(nodeZ);
	ch++;
}